### PR TITLE
Reduce VirtualHost memory utilization by avoiding CatchAllVirtualCluster when not needed

### PR DESCRIPTION
**Pull request by yurykats**
Reduce VirtualHost memory utilization by avoiding CatchAllVirtualClusterinstance when not needed.

This is a step towards more memory efficient config data structures. Issue #24154.

Currently an instance of CatchAllVirtualCluster is always created for each vhost, but is only used when there are other virtual clusters (and none of them matches). If there are no virtual clusters to begin with, CatchAllVirtualCluster is a pure overhead.

This PR switches an instance to a unique_ptr and only creates one when there are virtual clusters present.

Removed vcluster stats from xfcc_integration_test, since the test does not setup vclusters and the stats are no longer created by default (they were never updated before). vcluster stats are tested in router_test, tag_extractor_impl_test and http_integration.

Signed-off-by: Yury Kats <ykats@google.com>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by adisuissa** [comment]
Assigning Stephan (and myself, on behalf of Yury's request)
/assign @zuercher

File: source/common/router/config_impl.h [hunk 1]
```diff
@@ -315,7 +315,7 @@ class VirtualHostImpl : public VirtualHost, Logger::Loggable<Logger::Id::router>
   const bool include_attempt_count_in_response_;
   absl::optional<envoy::config::route::v3::RetryPolicy> retry_policy_;
   absl::optional<envoy::config::route::v3::HedgePolicy> hedge_policy_;
-  const CatchAllVirtualCluster virtual_cluster_catch_all_;
+  std::unique_ptr<CatchAllVirtualCluster> virtual_cluster_catch_all_;
```
Comment (adisuissa) [inline source/common/router/config_impl.h:318]: Why was the const removed?

File: source/common/router/config_impl.cc [hunk 2]
```diff
@@ -1733,10 +1731,14 @@ VirtualHostImpl::VirtualHostImpl(
...
-        VirtualClusterEntry(virtual_cluster, *vcluster_scope_,
-                            factory_context.routerContext().virtualClusterStatNames()));
+  if (!virtual_host.virtual_clusters().empty()) {
+    for (const auto& virtual_cluster : virtual_host.virtual_clusters()) {
+      virtual_clusters_.push_back(
+          VirtualClusterEntry(virtual_cluster, *vcluster_scope_,
+                              factory_context.routerContext().virtualClusterStatNames()));
+    }
+    virtual_cluster_catch_all_ = std::make_unique<CatchAllVirtualCluster>(
+        *vcluster_scope_, factory_context.routerContext().virtualClusterStatNames());
```
Comment (adisuissa) [inline source/common/router/config_impl.cc:1741]: IIUC vcluster_scope_ is passed and is not const (i.e., its state may be updated in the `CatchAllVirtualCluster` c'tor). I'm not sure if ordering of operations is important, but I guess to make it consistent with the previous code, this line can be executed at the beginning of the if block.

**Review by adisuissa — COMMENTED** [review]
Thanks, left a couple of minor comments.

File: source/common/router/config_impl.h (hunk 1 above)
Comment (yurykats) [reply source/common/router/config_impl.h]: Done

File: source/common/router/config_impl.cc (hunk 2 above)
Comment (yurykats) [reply source/common/router/config_impl.cc]: Done

**Review by adisuissa — DISMISSED** [review]
LGTM, thanks!

**Review by adisuissa — COMMENTED** [review]
Can you please check CI, the failure of `//test/integration:xfcc_integration_test` seems consistent.

**Comment by yurykats** [comment]
Removed vcluster stats from xfcc_integration_test, since the test does not setup vclusters and the stats are no longer created by default (they were never updated before).
vcluster stats are tested in router_test, tag_extractor_impl_test and http_integration.

**Comment by adisuissa** [comment]
> Removed vcluster stats from xfcc_integration_test, since the test does not setup vclusters and the stats are no longer created by default (they were never updated before). vcluster stats are tested in router_test, tag_extractor_impl_test and http_integration.

I think that is a valid claim, but want to another input. @jmarantz IIUC this PR removes some stats that currently appear, and as such it is a use-facing change. Assuming that these stats are always 0 in this case because vhost isn't explicitly configured (@yurykats correct me if I'm wrong), is it ok to remove them? If so, other than updating the release notes, is there anything else to consider or update?

**Comment by jmarantz** [comment]
I don't have a strong opinion about this particular stat. I'm not sure what monitoring might break for various Envoy deployments.  I feel like @zuercher would be able to provide insight or redirect to someone else if needed.

**Comment by yurykats** [comment]
> We need to add a release note that describes that change, however.

Thanks for the review, added to release notes.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Comment by jmarantz** [comment]
Removing myself as my review is not needed on this PR and I didn't actually review it :)

**Review by zuercher — APPROVED** [review]

**Pull request merged**